plat/hisilicon/hikey960/include/platform_def.h

```diff
@@ -31,8 +31,10 @@
 #define PLAT_NUM_PWR_DOMAINS		(PLATFORM_CORE_COUNT + \
 					 PLATFORM_CLUSTER_COUNT + 1)
 
-#define PLAT_MAX_RET_STATE		1
-#define PLAT_MAX_OFF_STATE		2
+#define PLAT_MAX_RUN_STATE		0
+#define PLAT_MAX_STB_STATE		1
```

Comment:
Hi @soby-mathew, let me give some background introduction for this.

This patch is to following the PSCI spec  ARM DEN 0022D, chapter "Recommended StateID Encoding" [1]. So all the state ID definitions are exactly same with "Table 16 StateID sample encoding". We have some discussion in Linux kernel mailing list for this as well [2], from the discussion my understanding is we should follow up the "Recommended StateID Encoding" if possible. So this is why this patch defines the exactly same value with spec.

After some discussion with @sudeep-holla, seems to me, if the recommended state ID can meet platform requirement then we can simply to follow up it (like Hikey960), otherwise the platform can define their own stateID. IMHO, this will be easily to follow up for SoC vendors as well. 

With your suggestion, that means we can just use PSCI spec as reference and the platform can be very flexible to define its own state ID. If you and Sudeep both could agree for this, I will change the patch with your suggestions. @sudeep-holla, could you also help review for this?

Please freely correct me if I misunderstand anything.

[1] http://infocenter.arm.com/help/topic/com.arm.doc.den0022d/Power_State_Coordination_Interface_PDD_v1_1_DEN0022D.pdf
[2] http://archive.armlinux.org.uk/lurker/message/20171124.143947.379e9894.en.html

Reply:
Ok, if it is unlikely that you want to support more retention states for this platform and you don't plan to reuse the encoding for other related hikey platforms, then you can stick to the recommendation as in the spec.

The wfi() is usually the `standby` state and it seems Hikey960 does not implement any other `retention` state. In ARM-TF, we consider standby as also one of the retention state which means Hikey 960 only has one `retention` state. Hence you can define PLAT_MAX_RET_STATE as 1. 

Also, usually linux has an optimization to call wfi() directly instead of issuing PSCI CPU SUSPEND smc call for `standby` and it is better to expose the standby state as such rather than as a different retention state.

Reply:
Yes, AFAIK Hikey960 only supports only one retention states.

But on Hikey960, except WFI state it does has another state named "CPU_NAP" state, this state is corresponding to 'PLAT_MAX_RET_STATE' in ARM-TF. From my understanding, the WFI is a state for CPU internal clock gating, on the other hand, "CPU_NAP" state is external clock gating and this state is supported by MCU firmware.

For WFI state, just as you said kernel directly calls wfi() to enter it. But "CPU_NAP" state need ARM-TF to support it, so this is corresponding to the definition PLAT_MAX_RET_STATE.

As result, I think we can remove the definition PLAT_MAX_STB_STATE, and we keep the definition as below:
+#define PLAT_MAX_RUN_STATE		0
+#define PLAT_MAX_RET_STATE		2
+#define PLAT_MAX_OFF_STATE		3

How about you think for this?

Reply:
@Leo-Yan , Thanks for the clarification. I didn't realise the implementation in `hikey960_pwr_domain_standby` correponded to the `CPU_NAP` state.

Yes, if you have a CPU_NAP state, then PLAT_MAX_RET_STATE must be 2.  So effectively, the Hikey960 can be thought to have 2 retention states : WFI and CPU_NAP.

The hikey960_pwr_idle_states[] should list all the possible valid combinations of power states at each level. The State-IDs 0x1, 0x2, 0x3, 0x33 are the valid ones. Currently 0x1 is missing from the list. 

hikey960_pwr_domain_standby() must now check if State_ID is 0x1 or 0x2. If `0x2`, enter CPU_NAP, else if `0x1` enter WFI. Although currently linux will not call PSCI for WFI state, ARM-TF must list this as a valid State-ID.

Reply:
@Leo-Yan, Would you like to do the above suggestion? It is not blockng for this PR and can be done later.

Reply:
@soby-mathew, sure, will commit new patch in today later. Thanks for suggestion.

Reply:
@soby-mathew, @davidcunado-arm, have updated patch according to comment. FYI.